<a href="https://colab.research.google.com/github/poonamaswani/DataScienceAndAI/blob/main/CAM_DS_C101_Demo_6_2_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<br>

**First things first** - please go to 'File' and select 'Save a copy in Drive' so that you have your own version of this activity set up and ready to use.
Remember to update the portfolio index link to your own work once completed!

# Demonstration 6.2.3 Clustering evaluation methods

Follow the demonstration to calculate the Davies-Bouldin index in Python. In this video, you will learn how to calculate the DBI for 2, 3, and $k$ clusters.

## a. Davies-Bouldin index for 2 clusters

Let's assume we have two clusters, $C_1$ and $C_2$, with the following two-dimensional points:

- $C_1$: Points $P{11} = (1,2)$, $P{12} = (2,3)$, and $P{13} = (1,3)$.
- $C_2$: Points $P{21} = (5,6)$, $P{22} = (6,5)$, and $P{23} = (5,5)$.

Step 1: Calculate the centroids $A_1$ and $A_2$ for both clusters.

Step 2: Calculate $S_i$ for each cluster, the average Euclidean distance of points in the cluster to the centroid.

Step 3: Calculate $M_{i,j}$, the Euclidean distance between centroids $A_1$ and $A_2$.

Step 4: Compute $R_{i,j}$ for each pair of clusters.

Step 5: Determine $D_i$ for each cluster,

Based on the example with the provided cluster points:

1. **Centroids Calculation:**
   - Centroid of $C_1$, $A_1$, is approximately (1.33, 2.67).
   - Centroid of $C_2$, $A_2$, is approximately (5.33, 5.33).

2. **Internal Dispersion ($S_i$):**
   - $S_1$, the average Euclidean distance of points in $C_1$ to $A_1$, is approximately 0.654.
   - $S_2$, the average Euclidean distance of points in $C_2$ to $A_2$, is also approximately 0.654.

3. **Centroid Separation ($M_i,j$):**
   - $M_{1,2}$, the Euclidean distance between $A_1$ and $A_2$, is approximately 4.807.

4. **DBI Calculation ($R_{i,j}$ and $D_i$):**
   - $R_{1,2}$, which compares the internal dispersion of $C_1$ and $C_2$ against their separation, is approximately 0.272.
   - Since we have only two clusters, $D_1 = R_{1,2}$ and $D_2 = R_{1,2}$, both are approximately 0.272.

5. **Overall Davies-Bouldin Index:**
   - The DBI is the average of $D_1$ and $D_2$, which in this case is 0.272.

This low DBI value suggests that the clusters are well-separated and have low internal dispersion, indicating a good clustering result.

In [ ]:
# Import the necessary libraries.
import numpy as np
from scipy.spatial import distance

# Define the points for each cluster.
C1_points = np.array([(1, 2), (2, 3), (1, 3)])
C2_points = np.array([(5, 6), (6, 5), (5, 5)])

# Calculate the centroids A1 and A2.
A1 = np.mean(C1_points, axis=0)
A2 = np.mean(C2_points, axis=0)

# Calculate Si for each cluster (average Euclidean distance to the centroid).
S1 = np.mean([distance.euclidean(p, A1) for p in C1_points])
S2 = np.mean([distance.euclidean(p, A2) for p in C2_points])

# Calculate Mi,j (Euclidean distance between centroids).
M12 = distance.euclidean(A1, A2)

# Compute Ri,j for the pair of clusters.
R12 = (S1 + S2) / M12

# Determine Di for each cluster (here we only have one other cluster).
D1 = R12
D2 = R12

# Calculate the Davies-Bouldin Index (DBI).
DBI = np.mean([D1, D2])

(A1, A2, S1, S2, M12, R12, D1, D2, DBI)

(array([1.33333333, 2.66666667]),
 array([5.33333333, 5.33333333]),
 0.6540388352636305,
 0.6540388352636305,
 4.8074017006186525,
 0.2720966026947421,
 0.2720966026947421,
 0.2720966026947421,
 0.2720966026947421)

## b. Davies-Bouldin index for 3 clusters

With three clusters, let's examine the calculation steps:

1. **Centroids Calculation:**
   - Centroid of $C_1, A_1$, is approximately (1.33, 2.67).
   - Centroid of $C_2, A_2$, is approximately (5.33, 5.33).
   - Centroid of $C_3, A$3$, is approximately (8.33, 8.33).

2. **Internal Dispersion ($S_i$):**
   - $S_1$, $S_2$, and $S_3$ are the average Euclidean distances within clusters $C_1$, $C_2$, and $C_3$ to their centroids, respectively. All have a value of approximately 0.654.

3. **Centroid Separation (Mi,j):**
   - $M_{1,2}$, $M_{1,3}$, and $M_{2,3}$ are the Euclidean distances between the centroids of different pairs of clusters:
   - Between $C_1$ and $C_2$, $M_{1,2}$ is approximately 4.807

In [ ]:
# Define the points for three clusters now.
C1_points = np.array([(1, 2), (2, 3), (1, 3)])
C2_points = np.array([(5, 6), (6, 5), (5, 5)])
C3_points = np.array([(8, 9), (9, 8), (8, 8)])

# Calculate the centroids A1, A2, and A3 for all clusters.
A1 = np.mean(C1_points, axis=0)
A2 = np.mean(C2_points, axis=0)
A3 = np.mean(C3_points, axis=0)

# Calculate Si for each cluster (average Euclidean distance to the centroid).
S1 = np.mean([distance.euclidean(p, A1) for p in C1_points])
S2 = np.mean([distance.euclidean(p, A2) for p in C2_points])
S3 = np.mean([distance.euclidean(p, A3) for p in C3_points])

# Calculate Mi,j (Euclidean distance between centroids of different clusters).
M12 = distance.euclidean(A1, A2)
M13 = distance.euclidean(A1, A3)
M23 = distance.euclidean(A2, A3)

# Compute Ri,j for each pair of clusters.
R12 = (S1 + S2) / M12
R13 = (S1 + S3) / M13
R23 = (S2 + S3) / M23

# Determine Di for each cluster (maximum Rij for each cluster compared to all others).
D1 = max(R12, R13)
D2 = max(R12, R23)
D3 = max(R13, R23)

# Calculate the Davies-Bouldin Index (DBI) as the average of all Di values.
DBI = np.mean([D1, D2, D3])

# Collect all results in a dictionary for better readability.
results = {"A1": A1, "A2": A2, "A3": A3,
           "S1": S1, "S2": S2, "S3": S3,
           "M12": M12, "M13": M13, "M23": M23,
           "R12": R12, "R13": R13, "R23": R23,
           "D1": D1, "D2": D2, "D3": D3,
           "DBI": DBI}

results

{'A1': array([1.33333333, 2.66666667]),
 'A2': array([5.33333333, 5.33333333]),
 'A3': array([8.33333333, 8.33333333]),
 'S1': 0.6540388352636305,
 'S2': 0.6540388352636305,
 'S3': 0.6540388352636306,
 'M12': 4.8074017006186525,
 'M13': 9.006170724070865,
 'M23': 4.2426406871192865,
 'R12': 0.2720966026947421,
 'R13': 0.14524237998633,
 'R23': 0.30831686371617617,
 'D1': 0.2720966026947421,
 'D2': 0.30831686371617617,
 'D3': 0.30831686371617617,
 'DBI': 0.29624344337569813}

## c. Davies-Bouldin index for $k$ clusters

The function `calculate_dbi` computes the Davies-Bouldin Index (DBI) for $k$ clusters. Given the example with three clusters, the DBI is approximately 0.296, and here are the details:

- **Centroids** of the clusters are approximately:
  - $A_1 = [1.33, 2.67]$
  - $A_2 = [5.33, 5.33]$
  - $A_3 = [8.33, 8.33]$

- **Internal Dispersion ($S_i$)** for each cluster:
  - $S_1 \approx S_2 \approx S_3 \approx 0.654$

- **R values ($R_{i,j}$)**, which are the ratios of the sum of internal dispersions to the centroid separation for all pairs of clusters:
  - $R_{1,2} \approx 0.272$
  - $R_{1,3} \approx 0.145$
  - $R_{2,3} \approx 0.308$
  - And the same values for the reverse pairs since $R{i,j} = R{j,i}$

- **D values ($D_i$)**, the maximum $R_{i,j}$ value for each cluster:
  - $D_1 \approx 0.272$
  - $D_2 \approx 0.308$
  - $D_3 \approx 0.308$

The function can be applied to any number of clusters to determine their DBI, allowing for an assessment of clustering quality across different clustering solutions.

In [ ]:
def calculate_dbi(clusters):
    """
    Calculates the Davies-Bouldin Index for K clusters.

    Parameters:
    - clusters: A list of arrays, where each array represents the points in a cluster.

    Returns:
    - dbi: The Davies-Bouldin Index for the clusters.
    - details: A dictionary containing centroids, internal dispersions, and R values for all cluster pairs.
    """
    # Calculate the centroids.
    centroids = [np.mean(cluster, axis=0) for cluster in clusters]

    # Calculate Si for each cluster.
    dispersions = [np.mean([distance.euclidean(p, centroid) for p in cluster]) for cluster, centroid in zip(clusters, centroids)]

    # Calculate Mi,j for all cluster pairs and Ri,j values.
    R_values = {}
    for i, (centroid_i, dispersion_i) in enumerate(zip(centroids, dispersions)):
        for j, (centroid_j, dispersion_j) in enumerate(zip(centroids, dispersions)):
            if i != j:
                Mij = distance.euclidean(centroid_i, centroid_j)
                Rij = (dispersion_i + dispersion_j) / Mij
                R_values[(i, j)] = Rij

    # Determine Di for each cluster.
    D_values = [max(R_values[(i, j)] for j in range(len(clusters)) if i != j) for i in range(len(clusters))]

    # Calculate the Davies-Bouldin Index as the average of Di values.
    dbi = np.mean(D_values)

    # Collect all intermediate results for analysi.
    details = {"centroids": centroids, "dispersions": dispersions,
              "R_values": R_values, "D_values": D_values}

    return dbi, details

# Example usage with the previously defined clusters
dbi, details = calculate_dbi([C1_points, C2_points, C3_points])
dbi, details

(0.29624344337569813,
 {'centroids': [array([1.33333333, 2.66666667]),
   array([5.33333333, 5.33333333]),
   array([8.33333333, 8.33333333])],
  'dispersions': [0.6540388352636305, 0.6540388352636305, 0.6540388352636306],
  'R_values': {(0, 1): 0.2720966026947421,
   (0, 2): 0.14524237998633,
   (1, 0): 0.2720966026947421,
   (1, 2): 0.30831686371617617,
   (2, 0): 0.14524237998633,
   (2, 1): 0.30831686371617617},
  'D_values': [0.2720966026947421, 0.30831686371617617, 0.30831686371617617]})

# Key information
This demonstration illustrated how to calculate the DBI.

## Reflect
What are the pracitical applications of this technique?

> Select the pen from the toolbar to add your entry.